util/devel/check_paths

```diff
@@ -0,0 +1,176 @@
+references by recognizing certain strings as likely to be the prefix
+of a path in the Chapel repository, e.g., "doc/" or "modules/" or
+"test/".
+"""
+
+import os.path
+import argparse
+
+def check_path(p):
+    """returns True if p exists, False if it doesn't"""
+
+    # These are common written-out alternatives, not paths
+    if p == "compiler/runtime":
+        return True
+    if p == "compiler/module":
+        return True
+
+    # These show up in generated html files relative to a "modules"
+    # directory that isn't the top-level one we're going to look in.
+    if p == "modules/chpldoc.doc.html":
+        return True
+    if p == "modules/chpldoc.doc/Defined.html":
+        return True
+    # Or in generated .js
+    if p == "modules/chpldoc.doc" or p == "modules/chpldoc.doc/Defined":
+        return True
+    # jquery .js file means a different test/ directory too
+    if p == "test/unit/core.js":
+        return True
+
+    return os.path.isfile(p) or os.path.isdir(p)
+
+# Memoize will be more useful once the script runs on multiple files
+# in one go.
+# Cribbed from a stackoverflow answer
+class Memoize:
+    def __init__(self, f):
+        self.f = f
+        self.memo = {}
+    def __call__(self, *args):
+        if not args in self.memo:
+            self.memo[args] = self.f(*args)
+        return self.memo[args]
+
+@Memoize
+def check_token(t):
+    """If t "looks like" a path but doesn't exist, return False.
+    Otherwise, return True."""
+
+    # Skip everything that doesn't look like something
+    if  not (t.startswith("doc/") or
+             t.startswith("test/") or
+             t.startswith("compiler/") or
+             t.startswith("runtime/") or
+             t.startswith("modules/") or
+             t.startswith("spec/") or
+             t.startswith("util/") or
+             t.startswith("tools/") or
+             t.startswith("third-party/") or
+             t.startswith("CHPL_HOME/") or
+             t.startswith("//github.com/chapel-lang/chapel/blob/master/")):
+        return True
+
+    elif t.startswith("CHPL_HOME/"):
+        t = t[10:]
+    elif t.startswith("//github.com/chapel-lang/chapel/blob/master/"):
+        t = t[44:]
+
+    # People like to say: "For details, see the README...".
+    while t.endswith("."):
+        t = t[:-1]
+
+    # Don't worry if it was "CHPL_HOME/..."
+    if t == "":
+        return True
+
+    # "*": often see foo/bar/*/baz, or foo/bar/baz-*
+    # We could just chop everything off after the * for the first case,
+    # But we'd need to do something about the basename, or expand the
+    # * for the second.  So just punt for now.
+    if "*" in t:
+        return True
+
+    firstcheck = check_path(t)
+
+    # Hack to check e.g. "util/files.cpp" against "compiler/utils/files.cpp"
+    if not firstcheck and t.startswith("util/"):
+        return check_path("compiler/" + t)
+
+    return firstcheck
+
+def special_exception(filename, token):
```

Comment:
Following up on top-level comment, a future addition could be to have exceptions saved to a file that is read in, rather than hard-coded, similar to `chplspell`.

Reply:
I'm leaving it a little loose here because the special exeptions aren't all substring checks, or all exact string equality checks, or all prefix checks, etc.

Reply:
Oh, the above comment should have been a response to the later comment about line 94.

For this section, I didn't want to come up with a file format for specifying ``<source file>:<allowed fake filename>`` pairs.

If the list of exceptions needs to change often, it may be worth keeping them in separate files.